### ISSUE TITLE
6-dof support with Xreal Air 2 Ultra

### ISSUE THREAD

**Issue by Kabouik** [iVideoGameBoss/PhoenixHeadTracker#18]
Xreal Air 2 Ultra support 6-dof, and combined with PhoenixHeadTeacker, it could finally make flight sims with those glasses an experience similar to full-fledged VR headsets. Do you plan on adding support for that in the future?

They even support hand-tracking, but I assume this would require much more development than just adding the extra sensors data for 6-dof.

**Comment by Droninator** [comment]
I would love this too, I waited for the ultra thinking it was compatible. If you need data or some with an ultra to work with I'd be open to helping in any way I can.

**Comment by Kabouik** [comment]
Same. Happy to help and provide data. I have Ultras and would love to use them to their full potential in flight sims, which means using PhoenixHeadTracker.

**Comment by pearce31** [comment]
Just got the ultras be happy to contribute

**Comment by Kabouik** [comment]
I tried using it with my Ultras and unfortunately PhoenixHeadTracker sees no data when I connect. I tried with original Airs too and it worked just fine, so there's clearly something Ultras-related.

Is there anything we can try @iVideoGameBoss? Maybe it's just that the sensors are not named the same way in the Ultras?

**Comment by Kabouik** [comment]
I guess we should do this: https://github.com/iVideoGameBoss/PhoenixHeadTracker?tab=readme-ov-file#you-can-use-phoenix-head-tracker-with-gyro-data-from-other-devices

I have zero experience in building stuff for Windows, but hopefully I can understand how it works later.

**Comment by Slion** [comment]
They added support for the Air 2 Ultra a couple of weeks ago so maybe it is just a simple matter of updating the DLL.
https://github.com/MSmithDev/AirAPI_Windows/releases/tag/v0.2

**Comment by Kabouik** [comment]
Yes it is. I've been talking with [@kawaharas](https://github.com/kawaharas) (who submitted the PR) and [@MSmithDev](https://github.com/MSmithDev) before the PR was merged and I made a fork of PhoenixHeadTracker at the time to try myself, and as far as I remember, the only thing that I needed to change to use it with my Ultra was the dll file. I didn't post the fork here because it was for my personal use until kawaharas' PR would be merged, but if it is of any use until @iVideoGameBoss includes the new dll here, it's available [here](https://github.com/Kabouik/PhoenixHeadTracker/archive/refs/heads/ultra.zip) for anyone who wants to use it temporarily.

I just updated it a minute ago because it was not using the latest dll release, but a WIP version that only supported the Ultra. Should be fine now. Please note that I have no skills for developping a Windows app, so this is not a fork trying to be an alternative to iVideoGameBoss', it's just to provide the executable and dll in one place until this one is updated.

**Comment by iVideoGameBoss** [comment]
I updated dll . Thank you.

**Comment by Kabouik** [comment]
Awesome, thanks. 

Back to the original issue, you mention in the Readme that it might be possible to add code to support 6-dof data. Would that require development on both PhoenixHeadTracker and the dll, or just PhoenixHeadTracker now that the dll supports the Ultra (I don't know if it collects data for 6-dof)? In case it's the latter, could you give any pointers?

**Comment by iVideoGameBoss** [comment]
The issue is that Xreal isn't providing PC developers access to their SDK for creating apps like PhoenixHeadTracker, as they’re prioritizing their Beam device sales. However, the [SDK](https://docs.xreal.com/) can still be used with the Unity Game Engine, though it appears to work only when running on an Android device. While it might be possible to use their Unity demos to communicate with OpenTrack via UDP without relying on AirAPI_Windows.dll, the SDK itself contains proprietary methods for handling 6DOF and drift correction. Unfortunately, this functionality is encrypted, and attempts to disassemble the SDK DLLs reveal no insights into how it works.

The AirAPI_Windows.dll identifies connected Xreal glasses by their PID (Product ID) and extracts raw sensor data. This data is then interpreted to estimate rotational values, converting them into degrees. While these calculations are not based on official specifications and are essentially an educated guess, they are sufficient for practical use.

For 6DOF functionality, the AirAPI_Windows.dll would need to process the raw data from the PID, making its best approximation of the 6DOF tracking data which would include and positional values. PhoenixHeadTracker could then utilize this approximate data to enable 6DOF head tracking.

The SDK in Unity can already handle 6DOF, but it’s mainly designed to work with Android devices. There are also projects on GitHub that connect to Xreal glasses to emulate virtual screens, but these projects typically don’t support 6DOF. If you want to implement 6DOF in a different environment, it may require some creative solutions or AI tools to extract and process the necessary data.

PhoenixHeadTracker is open-source, so if you’re considering creating a similar version or building one using Unity, leveraging AI tools could significantly streamline the process. Consider utilizing Google Gemini 2.0 with its Flash experimental model for advanced capabilities. Additionally, explore Claude.ai and DeepSeek R1 for complementary insights.

To start, you’ll need to provide these AI tools with the code for AirAPI_Windows.dll and instruct them to analyze and extract 6DOF data from the PID. This approach can help automate and refine the process of interpreting raw data for your application.

Just keep in mind that Xreal is focused on selling the Beam, so if they see you enabling 6DOF for PC gaming without their approval, they could shut down your project. Here’s an interesting tidbit: the Nebula app for PC still doesn’t work right, which speaks volumes about Xreal’s true level of commitment to PC support.

A complete 6DOF (Six Degrees of Freedom) system requires six values that describe an object's position and orientation in 3D space:
Positional Data (Translation)

These represent the object's position in 3D space along the X, Y, and Z axes:

    X (left-right): Horizontal movement, such as shifting left or right.
    Y (up-down): Vertical movement, such as rising or falling.
    Z (forward-backward): Depth movement, such as moving closer to or farther from the viewpoint.

Rotational Data (Orientation)

These define the object's orientation in terms of rotation around the three axes:
4. Yaw (rotation around the Y-axis): Turning left or right, like shaking your head "no."
5. Pitch (rotation around the X-axis): Tilting up or down, like nodding your head "yes."
6. Roll (rotation around the Z-axis): Tilting side to side, like tipping your head toward your shoulder.
Role of Accelerometer Data

While 6DOF tracking requires the above six values, accelerometers are essential for determining some of them indirectly:

    Motion Detection: Accelerometers measure acceleration along the X, Y, and Z axes, detecting changes in velocity and movement.
    Positional Estimation: By integrating acceleration over time, approximations of positional changes (X, Y, Z) can be derived.
    Tilt and Gravity Sensing: Accelerometers detect the gravity vector, providing critical input for determining pitch and roll.

Although accelerometers are vital, they are typically combined with gyroscopes and other sensors to refine 6DOF tracking and minimize drift or inaccuracies.

**Comment by MSmithDev** [comment]
@iVideoGameBoss 
Depending how much will power you have. It's possible to use the lib_nrapi.dll that comes in nebula for windows and make function calls to it. I started messing around with it by exploring the C# wrappers in the NRSDK and making similar calls to the dll. I didn't spend a ton of time on it yet but I was able to control thinks like the display mode (2D/3D) and get raw imu data. I started working on getting it to give me a stabilized head pose from the dll (which shouldn't have drift) but haven't spend enough time skimming the the NRSDK to figure out the whole process to do this. It seems like it would be possible. I also reached out to an xreal engineer to see if they could describe a process for this but haven't heard back yet.

**Comment by Slion** [comment]
Thanks guys for sharing your knowledge with us.

My understanding is that the IMU sensors are mostly used to compute the rotational data and thus enable 3DoF while the stereo cameras are used for positional data and give you 6DoF. So if you want 6DoF on PC you would need to get access to the cameras frames. I noticed that when I plug the Ultra into my PC it detects an UVC camera. However so far I was not able to get the frames through OpenCV.
